### ISSUE TITLE
Installed scripts can't know their own name

### ISSUE THREAD

**Issue by umbopepato** [denoland/deno#5725]
Since the removal of `window.location` (https://github.com/denoland/deno/pull/5034) there's no way for a `deno install`ed script to know its own name.

```
$ deno install -n cmdname https://url.to/script.ts
                  ^^^^^^^
```
This has a significant impact on CLI modules, that may need it for various reasons such as showing help messages and usage examples, registering shell completions etc.

Are there any plans to reintroduce this functionality somehow?

**Comment by nayeemrmn** [comment]
`window.location` wouldn't have shown you `cmdname`, it would have been `https://url.to/script.ts`.

**Comment by umbopepato** [comment]
That's right @nayeemrmn, I updated the issue.

**Comment by nayeemrmn** [comment]
Well if the use case is okay with everyone then a solution would be to inject it into the runtime in a special env var set by the auto-generated script:
```shell
#!/bin/sh
# generated by deno install
export DENO_INSTALLED_PATH=$0
deno "run" "https://url.to/script.ts" "$@"
```
```js
// namespace Deno {
//   installedPath: string | null;
// }
console.log(Deno.installedPath); // "/home/nayeem/.deno/bin/cmdname"
```

**Comment by c4spar** [comment]
It would be also nice, if there is a way to get the file name from a script which isn't installed with `deno install` but executed with `deno run`, without the need to set an environment variable or a cli flag.

**Comment by nayeemrmn** [comment]
@c4spar Getting the URL of the main module is an unrelated issue.

Note that a main module is already able to get its own name through `import.meta.url` and then pass it around. Adding a built-in global like `Deno.mainModule` would honestly promote a common bad practice... I don't think modules should be written to depend on this state.

**Comment by c4spar** [comment]
> @c4spar Getting the URL of the main module is an unrelated issue.
> 
> Note that a main module is already able to get its own name through `import.meta.url` and then pass it around. Adding a built-in global like `Deno.mainModule` would honestly promote a common bad practice... I don't think modules should be written to depend on this state.

Hey @nayeemrmn, thank you for answering me!
I know about `import.meta.url`. But i hoped for a single solution which works with both, `deno install` and `deno run`, without the need to pass `import.meta.url` around and checking for an environment variable which would me also force to run deno with the `--allow-env` flag.

**Comment by nayeemrmn** [comment]
@c4spar Again, there is a big difference between the script name as described in this issue and the main module URL you're talking about. It has nothing to do with `deno install` because it's defined the same either way. Secondly, the environment variable suggestion that I quickly decided wasn't necessary was for internal use, it wouldn't need `--allow-env`. Thirdly, I never suggested any of this for what you're asking for, which is unrelated to this issue.

**Comment by sholladay** [comment]
I think it would be nice if the script name were available via `Deno.scriptName` or similar. Would also be useful to have `Deno.rawArgs` which would include the full command used to execute the script.

**Comment by jsejcksn** [comment]
Consideration: What are the security and privacy implications of these kinds of introspection?

**Comment by umbopepato** [comment]
@jsejcksn one could be that if a script knows the location of its launcher it can grant itself more permissions by rewriting it (only requires `--allow-write`). And guessing the full path of the launcher from the script name, if installed in the default `--root`, is not that difficult knowing some env variables like `DENO_INSTALL` or `HOME`.

Don't know how viable it is, but maybe a possible security measure could be to exclude `bin` folders/files from write permissions unless explicitly listed in the flag?

```bash
# Can write everywhere except to path/to/.deno/bin/
$ deno run --allow-write myscript.ts

# Can write to path/to/.deno/bin/
$ deno run --allow-write=path/to/.deno/bin/ myscript.ts
```

Or maybe a file system permission wall?

Also, this may already be a concern since many cli script authors recommend an install name on their documentation so in most cases they already have all the info to infer the installation path.

**Comment by noahbrenner** [comment]
Of course, if a script can *read* the filesystem, it can probably find the script location (even if inefficiently) by `grep`ing for a unique string the script contains.

For the CLI use case, the most valuable information would just be the installed executable name itself. I don't imagine the full path to the script would be needed most of the time.

**Comment by rivy** [comment]
> For the CLI use case, the most valuable information would just be the installed executable name itself. I don't imagine the full path to the script would be needed most of the time.

Actually the full path is used quite frequently for associated CLI/executable data such as config files, etc.

**Comment by bartlomieju** [comment]
A script can determine its own name from `process.argv[1]` (equivalently `Deno.mainModule` / `import.meta.url`), which holds the path of the entrypoint module:

```ts
import process from "node:process";
import path from "node:path";

console.log(path.basename(process.argv[1])); // e.g. "script.ts"
// or, without node:
console.log(new URL(Deno.mainModule).pathname.split("/").pop());
```

Caveat for remote scripts: when a script is installed/run from a URL (e.g. `https://url.to/script.ts`), `process.argv[1]` / `Deno.mainModule` is that URL, so the "name" you get is the script's filename from the URL path rather than a local command name. `process.argv[0]` is always the `deno` executable, and `process.argv0` is `"deno"`, so neither of those helps here.

This mirrors how it works in Node: `process.argv[1]` carries the entrypoint path, and CLI libraries derive the program name from it. Closing as completed since the entrypoint name is available via `process.argv[1]`. If you specifically need the `deno install -n <name>` alias surfaced (distinct from the script filename), that's a separate enhancement and worth a focused issue.
